### PR TITLE
Blame Feature for vstest

### PR DESCRIPTION
**Pull request by vagisha-nidhi**
This feature gets the faulty test case name in event of a test host crash of reason unknown.
There are various scenarios in which the test host crashes due to a faulty test case and the reason cannot be diagnosed.
This pull request gets the name of the test case which caused the crash and displays it on the console.
To enable the blame mode, use --Blame|/Blame
 
The usage is as follows -
```
"vstest.console.exe" /Testadapterpath:"C:\TestProject\TestProject\bin\Debug\net46"  C:\path\to\your\assembly.dll  /Blame
```

Example : 
crash2

### CODE REVIEW

**Comment by harshjain2** [comment]
@dotnet-bot test Windows_NT / Debug Build

**Comment by vagisha-nidhi** [comment]
@dotnet-bot test Windows_NT / Debug Build

File: scripts/build.ps1 [hunk 1]
```diff
@@ -251,6 +251,7 @@ function Publish-Package
     $platformAbstractionNetCore = Join-Path $platformAbstraction $TPB_TargetFrameworkCore
     Copy-Item $platformAbstractionNet46\* $fullCLRPackageDir -Force
     Copy-Item $platformAbstractionNetCore\* $coreCLR20PackageDir -Force
+
```
Comment (singhsarab) [inline scripts/build.ps1]: Nit: remove this extra line.

File: scripts/build.ps1 [hunk 2]
```diff
@@ -270,6 +271,23 @@ function Publish-Package
...
     }
+
+    # Publish Datacollector
+    $blameDataCollector = Join-Path $env:TP_ROOT_DIR "src\Microsoft.TestPlatform.BlameDataCollector\bin\$TPB_Configuration"
+    $blameDataCollectorNet46 = Join-Path $blameDataCollector $TPB_TargetFramework
+    $blameDataCollectorNetCore = Join-Path $blameDataCollector $TPB_TargetFrameworkCore
+    Copy-Item $blameDataCollectorNet46\* $fullCLRPackageDir -Force
+    Copy-Item $blameDataCollectorNetCore\* $coreCLR20PackageDir -Force
+
+    # Note Note: Move the Blame data collector to extensions folder
```
Comment (singhsarab) [inline scripts/build.ps1]: Nit: Note appearing twice.
Why are copying and then moving them, why not directly copy to the required location?

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Xml;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    [DataCollectorFriendlyName("Blame")]
+    [DataCollectorTypeUri("my://blame/datacollector")]
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs]: Please change Uri to something like datacollector://microsoft/blame/1.0

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        public BlameCollector(IBlameFileManager blameFileManager)
+        {
+            this.blameFileManager = blameFileManager;
+        }
+
+        /// <summary>
+        /// Gets environment variables that should be set in the test execution environment
+        /// </summary>
+        /// <returns>Environment variables that should be set in the test execution environment</returns>
+        public IEnumerable<KeyValuePair<string, string>> GetTestExecutionEnvironmentVariables()
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs]: Where is this being used ? Why cant I return an empty list?

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            this.events.SessionEnd += this.SessionEnded_Handler;
+            this.events.TestCaseStart += this.Events_TestCaseStart;
+        }
+
+        /// <summary>
+        /// Called when Test Case Start event is invoked 
+        /// </summary>
+        private void Events_TestCaseStart(object sender, TestCaseStartEventArgs e)
+        {
+            EqtTrace.Info(Constants.TestCaseStart);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs]: Nit: While Logging, ideally we should log the class and method from where we are logging.

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            TestCase testcase = new TestCase(e.TestElement.FullyQualifiedName, e.TestElement.ExecutorUri, e.TestElement.Source);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs]: Is this code safe?
What if one of the parameters is null? Make sure no exceptions get thrown here.
If there can be cases like that, we should use try catch.

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        /// <summary>
+        /// Called when Session End event is invoked 
+        /// </summary>
+        private void SessionEnded_Handler(object sender, SessionEndEventArgs args)
+        {
+            var filepath = Path.Combine(AppContext.BaseDirectory, Constants.AttachmentFileName);
+            EqtTrace.Info(Constants.TestSessionEnd);
+            this.dataWriter = new BlameDataReaderWriter(TestSequence, filepath, this.blameFileManager);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameCollector.cs]: Why is datawriter a member variable if this is the only place it is being used?

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameDataReaderWriter.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        /// Writes all tests in the test sequence to the file
+        /// </summary>
+        public void WriteTestsToFile()
+        {
+            // Initialize Helper
+            blameFileManager.InitializeHelper();
+
+            // Adds all tests to file
+            blameFileManager.AddTestsToFormat(this.TestSequence, this.filePath);
+
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameDataReaderWriter.cs]: nit: remove extra spaces

**Pull request closed without merging**